File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-6.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// We can use recursive dfs where as we traverse each node in the tree
// we keep track of the maxValue seen so far
// if the curr node val is greater than or equal to maxVal so far, increment a res var
// Time: O(n) since we visit every node at most once
// Space: O(n) for the recursion call stack

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root: TreeNode | null): number {
        let res = 0

        const dfs = (node: TreeNode, maxSoFar: number) => {
            if (!node) return 

            if (node.val >= maxSoFar) res += 1

            const maxVal = Math.max(maxSoFar, node.val)

            dfs(node.left, maxVal)
            dfs(node.right, maxVal)
        }

        dfs(root, -101)
        return res
    }
}

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-6.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        cache = [None] * len(cost)

        def dfs(i):
            if i >= len(cost):
                return 0

            if cache[i] != None:
                return cache[i]

            left = dfs(i + 1)
            right = dfs(i + 2)

            # print(cost[i], left, right, cost[i] + min(left, right))
            cache[i] = cost[i] + min(left, right)

            return cache[i]

        return min(dfs(0), dfs(1))

        # Time: 2^n
        # Space: 2^n
        # For the recursive call stack
        